From e96dd9f9e94b034c46410f812088cff0d229c261 Mon Sep 17 00:00:00 2001
From: Block_Mrlimr267 <113236317+erha134@users.noreply.github.com>
Date: Tue, 30 Apr 2024 20:08:27 +0800
Subject: [PATCH] fix imports of modern forge support


diff --git a/proxy/src/main/java/io/izzel/lightfall/forge/ModernForgeServerConnector.java b/proxy/src/main/java/io/izzel/lightfall/forge/ModernForgeServerConnector.java
index a396b9a1..85b159db 100644
--- a/proxy/src/main/java/io/izzel/lightfall/forge/ModernForgeServerConnector.java
+++ b/proxy/src/main/java/io/izzel/lightfall/forge/ModernForgeServerConnector.java
@@ -1,6 +1,7 @@
 package io.izzel.lightfall.forge;
 
 import com.google.common.base.Preconditions;
+import io.izzel.lightfall.api.event.ModernFMLHandshakeEvent;
 import lombok.Getter;
 import net.md_5.bungee.BungeeServerInfo;
 import net.md_5.bungee.ServerConnector;
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 7595e037..5ca29e25 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -4,6 +4,8 @@ import com.google.common.base.Preconditions;
 import com.google.common.collect.HashMultimap;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Multimap;
+import io.izzel.lightfall.forge.ModernForgeClientHandler;
+import io.izzel.lightfall.forge.ModernForgeServerConnector;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
-- 
2.34.1

